packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2022, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {useContext} from 'react'
+import {ServerContext} from '../contexts'
+
+/**
+ * Server context
+ * @typedef {Object} ServerContext
+ * @property {Object} req - Request object
+ * @property {Object} res - Response object
+ */
+
+/**
+ * @callback serverCtxCallback
+ * @param {ServerContext} serverContext
+ */
+
+/**
+ * Execute the given function within the server context of a server-rendered page
+ *
+ * @param {serverCtxCallback} fn - Function will be called when the server has received the request but the response is not sent yet
+ * @example
+ * useServerContext(({req, res}) => { res.status(404) })
+ */
+export const useServerContext = (fn) => {
+    const serverContext = useContext(ServerContext)
+    const isOnServer = Boolean(serverContext.req)
+
+    if (isOnServer) {
+        fn(serverContext)
+    }
+}
```

Comment:
About the hook not returning anything: yes, it behaves more like `useEffect`, and I want it to be unlike the typical kinds of hooks.

Typically, hooks would return something. And this return value can be used in the rendering of the component html. But in our case, we want to discourage it from being used in the html (to avoid the mismatched html between server and client rendered, as was [commented earlier](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/708#discussion_r961040733)).

That's why I prefer having a callback. It focuses you on the request and response within the callback function, and less on the component html.

Reply:
About serializing request/response: currently, with the hook approach, there is no mechanism to serialize the request/response. It's possible to do that with `getProps`, though.

So as a result, if we try to print out the request/response, it would only work on the server side. Then React would complain about the mismatched html between server and client rendered.

Reply:
👀 I'll take a look about integrating the `useServerContext` hook with `useQuery` in the other branch.

Reply:
Both are good points.. thanks for clarifying those. I could get on board with them. 

With regards to usage with `useQuery` the way of setting status codes/headers would be like below? : 

```
const Page = () => {
    const {data, error} = useQuery(['my-query'], async () => api.getProduct(1))
    
    useServerContext((res) => {
        res.statusCode = error ? 404 : 200
    })

    return (
        <Fragment>
            {error ? <ErrorComponent error={error}>} : <Product {...data} />
        </Fragment>
    )
}
```

Reply:
@bendvc yup, that's right.. just like that.

I have made a branch here that demonstrates that it is possible to use both `useServerContext` and `useQuery`: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/713#discussion_r965084561